pandas/core/internals/blocks.py

```diff
@@ -2248,14 +2255,17 @@ def _astype(self, dtype, **kwargs):
     def _can_hold_element(self, element):
         tipo = maybe_infer_dtype_type(element)
         if tipo is not None:
-            return tipo == _NS_DTYPE or tipo == np.int64
+            return is_dtype_equal(tipo, self.dtype)
+        elif element is NaT:
+            return True
         elif isinstance(element, datetime):
+            if self.is_datetimetz:
+                return tz_compare(element.tzinfo, self.dtype.tz)
             return element.tzinfo is None
-        elif is_integer(element):
+        elif is_integer(element) and not isinstance(element, np.timedelta64):
```

Comment:
so I would move this type of check to is_valid_nat_for_dtype to avoid placing logic in these routines (which ends up getting scattered).

Reply:
> (which ends up getting scattered).

Yah, hopefully-before-too-long-term i think the best solution is for all of these to dispatch to the TimedeltaArray/DatetimeArray implementations.

> move this type of check to is_valid_nat_for_dtype

I considered that, but the trouble is that we really _don't_ want `iNaT` in there, see #16674.  Maybe a stricter version of `is_integer`?

This is definitely awkward; a couple of weird numpy behaviors are teaming up to make this necessary.

1) `issubclass(np.timedelta64, np.integer)` returns True, which in turn makes `is_integer(np.timedelta64(0))` return True.
2) numpy's deprecation warning about element-wise comparison is popping up when we compare `np.timedelta64(0) == iNaT` despite the fact that both are scalars.

Reply:
>  Maybe a stricter version of is_integer?

Looking at other uses of is_integer in this module, most of them would want to exclude timedelta64 objects

Reply:
there's another place in this module where we already have the check `if is_integer(value) and not isinstance(value, np.timedelta64):`

Reply:
> Looking at other uses of is_integer in this module, most of them would want to exclude timedelta64 objects
 
Does the whole world break if you just add the exclusion of timedelta64 to `is_integer` (instead of making an `is_integer_strict`)?

Reply:
will try this out in a dedicated branch